webhooks/webhooks.go

```diff
@@ -5,186 +5,92 @@ import (
-	}
-
-	// RenewedSubscriptionNotification is sent when a subscription renew.
-	// https://dev.recurly.com/page/webhooks#section-renewed-subscription
-	RenewedSubscriptionNotification struct {
-		Account      Account              `xml:"account"`
-		Subscription recurly.Subscription `xml:"subscription"`
-	}
-
-	// ExpiredSubscriptionNotification is sent when a subscription is no longer valid.
-	// https://dev.recurly.com/v2.4/page/webhooks#section-expired-subscription
-	ExpiredSubscriptionNotification struct {
-		Account      Account              `xml:"account"`
-		Subscription recurly.Subscription `xml:"subscription"`
+// parse parses an incoming webhook and returns the notification.
+func parse(r io.Reader, fn func(s string) (interface{}, error)) (interface{}, error) {
+	if closer, ok := r.(io.Closer); ok {
+		defer closer.Close()
 	}
 
-	// CanceledSubscriptionNotification is sent when a subscription is canceled.
-	// https://dev.recurly.com/page/webhooks#section-canceled-subscription
-	CanceledSubscriptionNotification struct {
-		Account      Account              `xml:"account"`
-		Subscription recurly.Subscription `xml:"subscription"`
+	notification, err := ioutil.ReadAll(r)
+	if err != nil {
+		return nil, err
 	}
-)
 
-// Invoice types.
-type (
-	// NewInvoiceNotification is sent when an invoice generated.
-	// https://dev.recurly.com/page/webhooks#section-new-invoice
-	NewInvoiceNotification struct {
-		Account Account `xml:"account"`
-		Invoice Invoice `xml:"invoice"`
+	var n notificationName
+	if err := xml.Unmarshal(notification, &n); err != nil {
+		return nil, err
 	}
 
-	// PastDueInvoiceNotification is sent when an invoice is past due.
-	// https://dev.recurly.com/v2.4/page/webhooks#section-past-due-invoice
-	PastDueInvoiceNotification struct {
-		Account Account `xml:"account"`
-		Invoice Invoice `xml:"invoice"`
+	dst, err := fn(n.XMLName.Local)
+	if err != nil {
+		return nil, err
 	}
-)
 
-// Payment types.
-type (
-	// SuccessfulPaymentNotification is sent when a payment is successful.
-	// https://dev.recurly.com/v2.4/page/webhooks#section-successful-payment
-	SuccessfulPaymentNotification struct {
-		Account     Account     `xml:"account"`
-		Transaction Transaction `xml:"transaction"`
+	if err := xml.Unmarshal(notification, dst); err != nil {
+		return nil, err
 	}
 
-	// FailedPaymentNotification is sent when a payment fails.
-	// https://dev.recurly.com/v2.4/page/webhooks#section-failed-payment
-	FailedPaymentNotification struct {
-		Account     Account     `xml:"account"`
-		Transaction Transaction `xml:"transaction"`
-	}
+	return dst, nil
+}
 
-	// VoidPaymentNotification is sent when a successful payment is voided.
-	// https://dev.recurly.com/page/webhooks#section-void-payment
-	VoidPaymentNotification struct {
-		Account     Account     `xml:"account"`
-		Transaction Transaction `xml:"transaction"`
+// dst returns the notification interface.
+func dst(s string) (dst interface{}, err error) {
```

Comment:
`dst` works well as a variable used close to unmarshaling etc, but as a package fn, even an unexported one, I'd like to be more explicit with naming. Instead of  `dst` / `origdst`, what about `nameToNotification` / `nameToNotificationDeprecated`?

Also `s string` to `name string`